fem/bilininteg_ext.cpp

```diff
@@ -359,19 +359,28 @@ static void OccaPADiffusionMultAdd3D(const int D1D,
 #define QUAD_2D_ID(X, Y) (X + ((Y) * Q1D))
 #define QUAD_3D_ID(X, Y, Z) (X + ((Y) * Q1D) + ((Z) * Q1D*Q1D))
 
-// PA Diffusion MultAdd 2D kernel
-template<const int D1D,
-         const int Q1D> static
-void PADiffusionMultAssembled2D(const int NE,
-                                const double* b,
-                                const double* g,
-                                const double* bt,
-                                const double* gt,
-                                const double* _op,
-                                const double* _x,
-                                double* _y)
+const int MAX_Q1D = 10;
+const int MAX_D1D = 10;
+
+// PA Diffusion Apply 2D kernel
+template<int T_D1D = 0, int T_Q1D = 0> static
+void PADiffusionApply2D(const int NE,
+                        const double* b,
+                        const double* g,
+                        const double* bt,
+                        const double* gt,
+                        const double* _op,
+                        const double* _x,
+                        double* _y,
+                        const int d1d = 0,
+                        const int q1d = 0)
 {
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
```

Comment:
I'm not sure that `D1D` and `Q1D` will be seen by the compiler as fixed (known at compile time) even when `T_D1D` and `T_Q1D` are nonzero. This can be tested by trying to create an object that uses them as template parameters, e.g. `TVector<D1D> test;`.

Reply:
I tried the above test and it seems that `D1D` is known at compile time since compilation completed successfully (I had to comment out the function instantiation with the default template parameters, of course).

This is a neat trick!

Reply:
It's indeed a nice trick!